core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -429,14 +433,123 @@ public void start(boolean isLeader) throws IOException {
         mStartupConsistencyCheck = getExecutorService().submit(new Callable<List<AlluxioURI>>() {
           @Override
           public List<AlluxioURI> call() throws Exception {
-            return checkConsistency(new AlluxioURI("/"), CheckConsistencyOptions.defaults());
+            return startupCheckConsistency(ExecutorServiceFactories
+                .fixedThreadPoolExecutorServiceFactory("startup-consistency-check", 32).create());
           }
         });
       }
     }
   }
 
   /**
+   * Checks the consistency of the root in a multi-threaded and incremental fashion.
+   *
+   * @return a list of paths in Alluxio which are not consistent with the under storage
+   * @throws InterruptedException if the thread is interrupted during execution
+   * @throws IOException if an error occurs interacting with the under storage
+   */
+  private List<AlluxioURI> startupCheckConsistency(final ExecutorService service)
+      throws InterruptedException, IOException {
+    /** A poison pill StartupConsistencyCheckers add to the queue to signal completion */
+    final long poison = -1;
```

Comment:
maybe call this `COMPLETED` or `COMPLETION_MARKER`? Based on its name it sounds like the everything should end when it is discovered, but it actually only increments the finished counter.